File: tree/606.根据二叉树创建字符串.cpp
/*
* @lc app=leetcode.cn id=606 lang=cpp
*
* [606] 根据二叉树创建字符串
*/

#include"head.h"
using namespace std;
// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
	string res;
public:
	void dfs(TreeNode* root) {
		if (!root) {
			res += "()";
			return;
		}
		//res+='(' + '0'+root->val;
		res.push_back('(');
		res+=to_string(root->val);
		bool a = root->left != NULL, b = root->right != NULL;
		if (a || b) {
			dfs(root->left);
			if (root->right)
				dfs(root->right);
		}
		res.push_back(')');
	}


	string tree2str(TreeNode* t) {
		dfs(t);
		return res.substr(1, res.size() - 2);
	}
};

// @lc code=end

File: string/606.根据二叉树创建字符串.go
/*
 * @lc app=leetcode.cn id=606 lang=golang
 *
 * [606] 根据二叉树创建字符串
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 var res string

func treeToStr(root *TreeNode) {
	if root == nil {
		res += "()"
		return
	}
	res += "(" + strconv.Itoa(root.Val)
	if root.Left == nil && root.Right == nil {
	} else if root.Left != nil && root.Right != nil {
		treeToStr(root.Left)
		treeToStr(root.Right)
	} else if root.Right == nil {
		treeToStr(root.Left)
	} else {
		res += "()"
		treeToStr(root.Right)
	}
	res += ")"
}

func tree2str(root *TreeNode) string {
	if root == nil {
		return "()"
	}
	res = ""
	treeToStr(root)
	return res[1 : len(res)-1]
}


// @lc code=end

